File: Programmers/LV3/43238.cpp
// 이분탐색
// 입국심사

#include <vector>
#include <algorithm>

using namespace std;

long long solution(int n, vector<int> times) {
    long long answer = 0;
    long long left = 0;
    long long right = (long long)(*max_element(times.begin(), times.end())) * n;
    while (left <= right) {
        long long middle = (left + right) / 2;
        long long result = 0;
        for (int time : times)
            result += middle / time;

        if (result >= n) {
            right = middle - 1;
            answer = middle;
        }
        else {
            left = middle + 1;
        }
    }
    return answer;
}

File: Programmers-py/43238.py
def solution(n, times):
    answer = 0
    time_left = 0
    time_right = max(times) * n
    while time_left <= time_right:
        cur_time = (time_left + time_right) // 2

        count_people = 0
        for time in times:
            count_people += cur_time // time

        if count_people >= n:
            time_right = cur_time - 1
            answer = cur_time
        else:
            time_left = cur_time + 1

    return answer